src/core/coap/coap.cpp

```diff
@@ -456,15 +628,465 @@ void CoapBase::DequeueMessage(Message &aMessage)
+        }
+
+        // Copy option
+        iterator.GetOptionValue(optionBuf);
+        SuccessOrExit(error = request->AppendOption(option->mNumber, option->mLength, optionBuf));
+
+        if (optionBuf != NULL)
+        {
+            free(optionBuf);
+            optionBuf = NULL;
+        }
+    }
+
+    if (!isOptionSet)
+    {
+        // Write Block2 option to next message
+        SuccessOrExit(error =
+                          request->AppendBlockOption(Message::kBlockType2, aMessage.GetBlockWiseBlockNumber() + 1,
+                                                     aMessage.IsMoreBlocksFlagSet(), aMessage.GetBlockWiseBlockSize()));
+        request->SetBlockWiseBlockNumber(aMessage.GetBlockWiseBlockNumber() + 1);
+        request->SetBlockWiseBlockSize(aMessage.GetBlockWiseBlockSize());
+        request->SetMoreBlocksFlag(aMessage.IsMoreBlocksFlagSet());
+
+        isOptionSet = true;
+    }
+
+    if (!aBeginBlock1Transfer)
+    {
+        DequeueMessage(aRequest);
+    }
+
+    otLogInfoCoap("Request Block2 Nr. %d, Size: %d bytes", request->GetBlockWiseBlockNumber(),
+                  otCoapBlockSizeFromExponent(request->GetBlockWiseBlockSize()));
+
+    SuccessOrExit(error =
+                      SendMessage(*request, aMessageInfo, TxParameters::GetDefault(), aCoapMetadata.mResponseHandler,
+                                  aCoapMetadata.mResponseContext, NULL, aCoapMetadata.mBlockwiseReceiveHook));
+
+exit:
+    if (error != OT_ERROR_NONE && request != NULL)
+    {
+        request->Free();
+    }
+
+    return error;
+}
+
+otError CoapBase::ProcessBlock1Request(Message &               aMessage,
+                                       const Ip6::MessageInfo &aMessageInfo,
+                                       const Resource &        aResource,
+                                       uint32_t                aTotalLength)
+{
+    otError  error                                        = OT_ERROR_NONE;
+    Message *response                                     = NULL;
+    uint8_t  buf[OPENTHREAD_CONFIG_COAP_MAX_BLOCK_LENGTH] = {0};
+    uint16_t bufLen                                       = OPENTHREAD_CONFIG_COAP_MAX_BLOCK_LENGTH;
+
+    SuccessOrExit(error = aMessage.ReadBlockOptionValues(OT_COAP_OPTION_BLOCK1));
+
+    // Read and then forward payload to receive hook function
+    bufLen = aMessage.Read(aMessage.GetOffset(), aMessage.GetLength() - aMessage.GetOffset(), buf);
+    SuccessOrExit(error = aResource.mReceiveHook(aResource.mContext, buf,
+                                                 otCoapBlockSizeFromExponent(aMessage.GetBlockWiseBlockSize()) *
+                                                     aMessage.GetBlockWiseBlockNumber(),
+                                                 bufLen, aMessage.IsMoreBlocksFlagSet(), aTotalLength));
+
+    if (aMessage.IsMoreBlocksFlagSet())
+    {
+        // Set up next response
+        VerifyOrExit((response = NewMessage()) != NULL, error = OT_ERROR_FAILED);
```

Comment:
Is `OT_ERROR_NO_BUFS` more appropriate here?
```suggestion
        VerifyOrExit((response = NewMessage()) != NULL, error = OT_ERROR_NO_BUFS);
```